abi.md

```diff
@@ -0,0 +1,147 @@
+title: Specification for the FIRRTL ABI
+revisionHistoryAbi: true
+---
+
+# Introduction
+
+FIRRTL defines a language/IR for describing synchronous hardware circuits.  This
+document specifies the mapping of FIRRTL constructs to Verilog in a manner
+similar to an application binary interface (ABI) which enables predictability of
+the output of key constructs necessary for the interoperability between circuits
+described in FIRRTL and between other languages and FIRRTL output.
+
+This document describes multiple versions of the ABI, specifically calling
+specific changes in later versions.  It is expected that a conforming FIRRTL
+compiler can lower to all specified ABIs.  This mechanism exists to allow
+improved representations when using tools which have better Verilog support and
+allow migration of existing development flows to significant representational
+changes.
+
+# FIRRTL System Verilog Interface
+
+To use a circuit describe in FIRRTL in a predictable way, the mapping of certain
+behaviors and boundary constructs in FIRRTL to System Verilog must be defined.
+Where possible this ABI does not impose constraints on implementation,
+concerning itself primarily with the boundaries of the circuit.
+
+## On Modules
+
+### The Circuit and Top Level
+
+The top level module, as specified by the circuit name, shall be present as a
+System Verilog module of the same name.  This firrtl module is considered a
+"public" module and subject to the lowering constraints for public modules.
+
+### External Modules
+
+An external module may be presumed to exist following the lowering constraints
+for public modules.  The module shall exist with a verilog name matching the 
+defname value or, lacking that, the module name.
+
+###  Public Modules
+
+Any module considered a "public" module shall be implemented in Verilog in a
+consistent way.  Any public module shall exist as a Verilog module of the same
+name.
+
+Each public module with definitions (e.g. not external modules) shall be placed
+in a file with the same name.  No assumption is made of the filename of an
+implementation of an external module, it is the user's job to include files in
+such a way in their tools to resolve the name.
+
+### Port Lowering ABIv1
+
+Ports are generally lowered to netlist types, except where Verilog's type system
+prevents it.
+
+Ports of integer types shall be lowered to netlist ports (`wire`{.verilog}) as a
+packed vector of equivalent size.  For example, consider the following FIRRTL:
+
+```firrtl
+circuit Top :
+  module Top :
+    output out: UInt<16>
+    input b: UInt<32>
+```
+
+This is lowered to the following Verilog:
+
+```verilog
+module Top(
+    output wire [15:0] out,
+    input wire [31:0] in
+);
+```
+
+Ports of aggregate type shall be lowered according to the "Aggregate Type
+Lowering" description in the firrtl spec.
+
+Ports of enum type shall be lowered to ports of a type consistent with the rules
+for lowering firrtl enum types to Verilog.
```

Comment:
This seems like a punt. Could we either be explicit about what we mean by "rules for lowering firrtl enum types" or else make a note admitting that we aren't ready to commit to a single rule at this point in time?

Reply:
put the forward reference in.  Will add the lowering.